tests/Adapter/WorseReflection/WorseBuilderFactoryTest.php

```diff
@@ -145,18 +202,72 @@ public function testStaticMethod()
         $this->assertTrue($source->classes()->first()->methods()->first()->isStatic());
     }
 
+    public function testTraitMethod()
+    {
+        $source = $this->build('<?php trait Foobar { public function method() {} }');
+        $this->assertEquals('method', $source->traits()->first()->methods()->first()->name());
+    }
+
+    public function testTraitMethodWithReturnType()
+    {
+        $source = $this->build('<?php trait Foobar { public function method(): string {} }');
+        $this->assertEquals('string', $source->traits()->first()->methods()->first()->returnType());
+    }
+
+    public function testTraitMethodProtected()
+    {
+        $source = $this->build('<?php trait Foobar { protected function method() {} }');
+        $this->assertEquals('protected', $source->traits()->first()->methods()->first()->visibility());
+    }
+
+    public function testTraitMethodWithParameter()
+    {
+        $source = $this->build('<?php trait Foobar { public function method($param) {} }');
+        $this->assertEquals('param', $source->traits()->first()->methods()->first()->parameters()->first()->name());
+    }
+
+    public function testTraitMethodWithParameterByReference()
+    {
+        $source = $this->build('<?php trait Foobar { public function method(&$param) {} }');
+        $this->assertTrue($source->traits()->first()->methods()->first()->parameters()->first()->byReference());
+    }
+
+    public function testTraitMethodWithTypedParameter()
+    {
+        $source = $this->build('<?php trait Foobar { public function method(string $param) {} }');
+        $this->assertEquals('string', (string) $source->traits()->first()->methods()->first()->parameters()->first()->type());
+    }
+
+    public function testTraitMethodWithAliasedParameter()
+    {
+        $source = $this->build('<?php use Foobar as Barfoo; trait Foobar { public function method(Barfoo $param) {} }');
+        $this->assertEquals('Barfoo', (string) $source->traits()->first()->methods()->first()->parameters()->first()->type());
+    }
+
+    public function testTraitMethodWithDefaultValue()
+    {
+        $source = $this->build('<?php trait Foobar { public function method($param = 1234) {} }');
+        $this->assertEquals(1234, (string) $source->traits()->first()->methods()->first()->parameters()->first()->defaultValue()->value());
+    }
+
+    public function testTraitMethodWithDefaultValueQuoted()
+    {
+        $source = $this->build('<?php trait Foobar { public function method($param = "1234") {} }');
+        $this->assertEquals('1234', (string) $source->traits()->first()->methods()->first()->parameters()->first()->defaultValue()->value());
+    }
```

Comment:
Are these property / method tests necessary? These things are already tested individually, I guess the Trait logic doesn't affect them, in which case we don't benefit from duplicating the tests?

Reply:
It's possible, but honestly I'm not certain: I'm still very new to testing overall and before this PR I hadn't really dug into the internals of Phpactor at all. It wasn't until I added these extra tests that I found the commented `todo` block, so they definitely do _something_ - it's just a question of which ones are really necessary.

One thing I noticed with the property tests, the `...Typed` ones have a default value in the respective class versions. This had to be removed in the Trait copies to get them passing, but I put that down to 'skipping due to lack of default property value support](https://github.com/phpactor/code-builder/pull/9/files#diff-5ef7cbf371f45e47f8cab19dd507146dR114). Again I don't know enough to say for sure, but there certainly seems to be something different.

I guess they can't hurt as a "fireguard" in case the property/method handling ever changes? I'm happy to remove some though if duplication is still a concern.